dask_cuda/benchmarks/local_cudf_groupby.py

```diff
@@ -0,0 +1,273 @@
+    print_key_value(key="Tree-reduction width", value=f"{args.split_every}")
+    if args.device_memory_limit:
+        print_key_value(
+            key="Device memory limit", value=f"{format_bytes(args.device_memory_limit)}"
+        )
+    print_key_value(key="RMM Pool", value=f"{not args.disable_rmm_pool}")
+    if args.protocol == "ucx":
+        print_key_value(key="TCP", value=f"{args.enable_tcp_over_ucx}")
+        print_key_value(key="InfiniBand", value=f"{args.enable_infiniband}")
+        print_key_value(key="NVLink", value=f"{args.enable_nvlink}")
+    print_key_value(key="Worker thread(s)", value=f"{args.threads_per_worker}")
+    print_key_value(key="Data processed", value=f"{format_bytes(results[0][0])}")
+    print_key_value(key="Output size", value=f"{format_bytes(results[0][1])}")
+    if args.markdown:
+        print("\n```")
+    data_processed, output_size, durations = zip(*results)
+    print_throughput_bandwidth(
+        args, durations, data_processed, p2p_bw, address_to_index
+    )
+
+
+def create_tidy_results(args, p2p_bw, results):
+    configuration = {
+        "dataframe_type": "cudf" if args.type == "gpu" else "pandas",
+        "shuffle": args.shuffle,
+        "sort": args.sort,
+        "split_out": args.split_out,
+        "split_every": args.split_every,
+        "in_parts": args.in_parts,
+        "rows_per_chunk": args.chunk_size,
+        "unique_ratio": args.unique_ratio,
+        "protocol": args.protocol,
+        "devs": args.devs,
+        "device_memory_limit": args.device_memory_limit,
+        "rmm_pool": not args.disable_rmm_pool,
+        "tcp": args.enable_tcp_over_ucx,
+        "ib": args.enable_infiniband,
+        "nvlink": args.enable_nvlink,
+    }
+    timing_data = pd.DataFrame(
+        [
+            pd.Series(
+                data=ChainMap(
+                    configuration,
+                    {
+                        "wallclock": duration,
+                        "data_processed": data_processed,
+                        "output_size": output_size,
+                    },
+                )
+            )
+            for data_processed, output_size, duration in results
+        ]
+    )
+    return timing_data, p2p_bw
+
+
+def parse_args():
+    special_args = [
+        {
+            "name": "--in-parts",
+            "default": 100,
+            "metavar": "n",
+            "type": int,
+            "help": "Number of input partitions (default '100')",
+        },
+        {
+            "name": [
+                "-c",
+                "--chunk-size",
+            ],
+            "default": 1_000_000,
+            "metavar": "n",
+            "type": int,
+            "help": "Chunk size (default 1_000_000)",
+        },
+        {
+            "name": "--unique-ratio",
+            "default": 0.01,
+            "type": float,
+            "help": "Fraction of rows that are unique groups",
+        },
+        {
+            "name": "--sort",
```

Comment:
Note that I am seeing errors when I run the benchmark with `--shuffle "explicit-comms" --split_out 2 --sort`, and the cause seems to be that dask is using [a shallow copy](https://github.com/dask/dask/blob/f382d9e8439f5ed200da6ce66df7b2b33a0fc500/dask/dataframe/utils.py#L726) in [DataFrame.drop](https://github.com/dask/dask/blob/f382d9e8439f5ed200da6ce66df7b2b33a0fc500/dask/dataframe/core.py#L5304), and [this](https://github.com/rapidsai/cudf/blob/4023b65233d8ade1b19f883b1e48abf614813254/python/dask_cudf/dask_cudf/sorting.py#L293) `"_partitions"`-column `drop` is somehow not working right after an explicit-comms shuffle.

I'm not exactly sure why the `"_partitions"` column is not getting dropped properly here (I am struggling to create a simple-enough reproducer to properly explore). However, I suspect that the custom round-tripping of of the input dataframe from a persisted list of futures is somehow running us into an undefined edge case.

Reply:
I see a (possibly different) error:

```
python -m dask_cuda.benchmarks.local_cudf_groupby --shuffle explicit-comms  --split_out 2 --sort -d 0,1,3,4   -p tcp
=>   File "/root/dask/dask/dataframe/utils.py", line 429, in check_matching_columns
    raise ValueError(
ValueError: The columns in the computed data do not match the columns in the provided metadata
Order of columns does not match
```
With the columns being:
```
Index(['key', 'int64___count', 'int64___max', 'float64___count',
       'float64___sum'],
      dtype='object')

Index(['key', 'int64___max', 'int64___count', 'float64___sum',
       'float64___count'],
      dtype='object')
```

Reply:
>This occurs also with --shuffle tasks

Oh no! You see this same error for the command you shared above, but with `--shuffle tasks`?

Reply:
yes (not always), more likely with more devices

Reply:
Thanks for pointing this out @wence-.  It seems like we need to tweak the groupby-aggregation functions (defined in `dask_cudf`) to output intermediate `DataFrame` results with **sorted** column names. Otherwise the ordering is non-deterministic, and we open ourselves up to failed metadata-enforcement checks in down-stream operations :/

Reply:
**Update**: https://github.com/rapidsai/cudf/pull/11836 should resolve this problem in `dask_cudf`.